### PR TITLE
Manual browser parse IDL tests and adjustments to correct parse failures

### PR DESCRIPTION
**Pull request by m-cheung**
Implementing requested tests for #39 
Attempting to resolve the issues described in #48

### CODE REVIEW

File: lib/org/chromium/webidl/Parser.js [hunk 1]
```diff
@@ -172,7 +172,7 @@ foam.CLASS({
               InterfaceLike: tseq(
                   alt('interface', 'exception'),
                   sym('identifier'), sym('Inheritance'), '{',
-                  sym('InterfaceMembers'), '}', ';'),
+                  sym('InterfaceMembers'), '}', sym('SemiColon')),
```
Comment (m-cheung) [inline lib/org/chromium/webidl/Parser.js:175]: Rules had to be relaxed for Interface declarations as many found in WebKit IDL files do not end with a semicolon

File: lib/org/chromium/webidl/WebKitParser.js [hunk 2]
```diff
@@ -33,6 +33,9 @@ foam.CLASS({
                       repeat0(alt(notChars('*'), seq('*', notChars('\/')))),
                       '*\/'),
                   // C-style preprocessor directives.
+                  seq('#ifndef',
+                    repeat0(notChars('#')),
+                    '#endif'),
```
Comment (m-cheung) [inline lib/org/chromium/webidl/WebKitParser.js]: The current code only recognized one line C-style preprocessor directives. Adding support for `#ifndef` here as it shows up quite often. May be changed in the future to account for other multiline directives

File: lib/org/chromium/webidl/WebKitParser.js [hunk 3]
```diff
@@ -33,6 +33,9 @@ foam.CLASS({
                       repeat0(alt(notChars('*'), seq('*', notChars('\/')))),
                       '*\/'),
                   // C-style preprocessor directives.
+                  seq(alt('#ifdef', '#ifndef'),
```
Comment (mdittmer) [inline lib/org/chromium/webidl/WebKitParser.js]: Why is this change necessary?

File: lib/org/chromium/webidl/WebKitParser.js [hunk 4]
```diff
@@ -52,10 +55,23 @@ foam.CLASS({
                   sym('ExtendedAttributeNamedArgList'),
                   sym('ExtendedAttributeIdentifierOrString'),
                   sym('ExtendedAttributeArgList'),
+                  sym('ExtendedAttributeIdentifier'),
```
Comment (mdittmer) [inline lib/org/chromium/webidl/WebKitParser.js]: Doesn't this look like a function call, not an identifier? Maybe name it `ExtendedAttributeFnCall` or similar? Also, can we get a comment explaining why this is needed? I assume it appears, in practice, in WebKit's web platform IDL; maybe link to an example.

File: lib/org/chromium/webidl/WebKitParser.js [hunk 5]
```diff
@@ -52,10 +55,23 @@ foam.CLASS({
...
                   sym('ExtendedAttributeNoArgs')),
               ExtendedAttributeLogicalOperators: tseq(
                   sym('identifier'), '=', sym('identifier'), alt('&', '|'),
                   tplus(sym('identifier'), alt('&', '|'))),
+              ExtendedAttributeIdentifier: tseq(
+                sym('identifier'), '(', sym('identifier'), ')'),
+
+              /* There is one file within the WebKit repository that currently
```
Comment (mdittmer) [inline lib/org/chromium/webidl/WebKitParser.js]: To keep the file consistent, let's use

```js
// This style of
// comment.
```

File: test/node/blinkParseIDLFileDAO-manual.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+var gitilesBaseURL = 'https://chromium.googlesource.com/chromium/src/+';
+var config = {
+  description: 'Blink IDL file DAO Parser manual test',
+  repositoryURL: 'https://chromium.googlesource.com/chromium/src.git',
+  localRepositoryPath: require('path').resolve(__dirname, '../data/blink/git'),
+  sparsePath: 'third_party/WebKit/Source',
+  findExcludePatterns: ['*/testing/*', '*/bindings/tests/*', '*/mojo/*'],
+  extension: 'idl',
+  parser: 'Blink'
```
Comment (mdittmer) [inline test/node/blinkParseIDLFileDAO-manual.js]: nit: Include trailing comma.

File: test/node/geckoParseIDLFileDAO-manual.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+  description: 'Gecko IDL file DAO manual test',
+  repositoryURL: 'https://github.com/mozilla/gecko-dev.git',
+  localRepositoryPath: require('path').resolve(__dirname, '../data/gecko/git'),
+  sparsePath: 'dom',
+  findExcludePatterns: ['*/test/*'],
+  extension: 'webidl',
+  parser: 'GeckoParser'
+};
+config.idlFileContentsFactory = function(path, contents) {
+  // Classes are injected by manualLocalGitTest.
```
Comment (mdittmer) [inline test/node/geckoParseIDLFileDAO-manual.js]: This comment is relevant, but the name is wrong, I think. Shouldn't it be `parseIDLFileDAOTest`? It should probably appear in each `config.idlFileContentsFactory = function...` too, to avoid confusion.

File: test/node/parseIDLFileDAO-helper.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    var originalTimeout;
+
+    afterAll(function() {
+      execSync(`/bin/rm -rf "${data.localRepositoryPath}"`);
+    });
+
+    beforeEach(function() {
+      data.GitilesIDLFile = foam.lookup('org.chromium.webidl.GitilesIDLFile');
+      data.GithubIDLFile = foam.lookup('org.chromium.webidl.GithubIDLFile');
+      data.BlinkParser = foam.lookup('org.chromium.webidl.Parser');
```
Comment (mdittmer) [inline test/node/parseIDLFileDAO-helper.js]: We can probably do:

```js
data.Parser = foam.lookup(`org.chromium.webidl.${data.parser}`);
```
Then use `data.Parser` later.

We will have to update the Blink config to set `config.parser = 'Parser'`, and add a comment that Blink uses default parser.

File: lib/org/chromium/webidl/WebKitParser.js (hunk 3 above)
Comment (m-cheung) [reply lib/org/chromium/webidl/WebKitParser.js]: This change was made while we were still processing the files in `win/interfaces`. Our current implementation only ignores one line preprocessor directives (such as `#define`). 

I have verified that this is not needed when only parsing files in `WebCore`

File: lib/org/chromium/webidl/WebKitParser.js (hunk 4 above)
Comment (m-cheung) [reply lib/org/chromium/webidl/WebKitParser.js]: This is actually not required anymore either. This was necessary because files within `win/interfaces` had attributes such as `UUID(78f416b8-e30d-4886-9884-6ace2f7e21f7)`

None of the files in WebCore seem to have this

File: lib/org/chromium/webidl/WebKitParser.js [hunk 9]
```diff
@@ -56,6 +56,17 @@ foam.CLASS({
               ExtendedAttributeLogicalOperators: tseq(
                   sym('identifier'), '=', sym('identifier'), alt('&', '|'),
                   tplus(sym('identifier'), alt('&', '|'))),
+              ExtendedAttributeFnCall: tseq(
+                sym('identifier'), '(', sym('identifier'), ')'),
```
Comment (arobins) [inline lib/org/chromium/webidl/WebKitParser.js]: Should be +4 indent, not +2.

Comment (m-cheung) [reply lib/org/chromium/webidl/WebKitParser.js]: Whoops. This has actually been removed because the semantic rule no longer necessary (only needed for files within `win/interfaces`

**Comment by m-cheung** [comment]
All requested changes should be addressed now

File: test/node/webKitParseIDLFileDAO-manual.js [hunk 10]
```diff
@@ -14,6 +14,7 @@ var config = {
   parser: 'WebKitParser',
 };
 config.idlFileContentsFactory = function(path, contents) {
+  // Classes are injected by parseIDLFileDAOTest
```
Comment (mdittmer) [inline test/node/webKitParseIDLFileDAO-manual.js]: nit: End comment with period.

**Review by mdittmer — APPROVED** [review]
LGTM. One nit about comment style.

**Comment by mdittmer** [comment]
@m-cheung forgot to let you address nit before merging. Feel free to push comment change right to master.
